### PR TITLE
Refactor performance tests artifacts handling

### PR DESCRIPTION
**Pull request by WunderBart**
## What?

TL;DR: 
- [Save all the performance results](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Save%20all%20the%20performance%20result%20files%20to%20the%20main%20artifacts%20folder) to the main `./artifacts` folder regardless of the triggering environment:
   [CI via CLI](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/.github/workflows/performance.yml#L40) / [locally via CLI](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/bin/plugin/cli.js#L93-L112) / [locally via npm](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/package.json#L308), etc.
- In the CI, [archive all results](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Archive%20performance%20results%20for%20each%20comparison%20step) (raw + calculated) for all [comparison jobs](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/.github/workflows/performance.yml#L38-L81) instead of only raw ones for the `current-branch vs. trunk` comparison, as they provide useful information,
- [Sort results by the test name](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Sort%20results%20and%20test%20logs%20primarily%20by%20the%20suite%20name) instead of branch name to reflect the actual order they're run at,
- [Create the wp-env configs at runtime](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Write%20the%20env%2Dspecific%20configs%20directly) instead of copying and altering the [`.wp-env.performance.json`](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/bin/plugin/utils/.wp-env.performance.json) boilerplate.

### Save all the performance result files to the main artifacts folder

Instead of storing files around the source code, let's use the already available [`WP_ARTIFACTS_PATH` directory](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-scripts/#failed-test-artifacts).

The `process.env.WP_ARTIFACTS_PATH` has been primarily used for storing the failed test artifacts (screenshots, HTML snapshots), but we can also use it for storing any other artifacts, like performance results or temporary trace files. Those files are currently stored alongside the source code and then copied to an intermediate `./__test-results/` folder, which is then picked up by the results upload CI action. After this PR, those files will be saved directly to the `WP_ARTIFACTS_PATH` folder, so we won't need to:
- guess the paths,
- copy results saved alongside the relative test file, and
- create any extra folders to handle the artifacts upload.

Once created, they're ready to be picked up by the upload action because the CI will define the `WP_ARTIFACTS_PATH` variable. 

Additionally, the final (calculated) results file that we currently save to the [enigmatic path](https://github.com/WordPress/gutenberg/blob/trunk/bin/plugin/commands/performance.js#L484-L488) (before [uploading them](https://github.com/WordPress/gutenberg/blob/04b53b71e2111ab2426b3ee4a9a717a62337afa7/bin/log-performance-results.js) to [codehealth](https://codehealth.vercel.app/project/gutenberg)), will now be saved to the `WP_ARTIFACTS_PATH` as well and zipped with the other (raw) results.

Finally, if the CI job fails, the results upload step will not occur. However, if there are any results that have already been created, they will still be uploaded along with the failure artifacts.

> **Note**
> The [performance runner script](https://github.com/WordPress/gutenberg/blob/04b53b71e2111ab2426b3ee4a9a717a62337afa7/bin/plugin/commands/performance.js) spawns the branch jobs in separate child processes, so they don't have access to the main `process.env` object provided by the CI. To solve that, I'm passing the relevant env variables to the child processes via [the `runShellScript` util](https://github.com/WordPress/gutenberg/blob/b9ac9c3f6dcd98ab08b129214d82c9b8744fc627/bin/plugin/lib/utils.js#L37).

### Archive performance results for each [comparison step](https://github.com/WordPress/gutenberg/blob/04b53b71e2111ab2426b3ee4a9a717a62337afa7/.github/workflows/performance.yml#L54-L90)

Currently, we are archiving performance results only for the `Compare performance with trunk` step. Now, the results will be uploaded for every comparison step as they provide useful information:

- Compare performance with trunk
- Compare performance with current WordPress Core and previous Gutenberg versions
- Compare performance with base branch
- Compare performance with custom branches

In case of a failure, the results upload action will not be dispatched. If any result files are written during a failed job, though, they will be uploaded within the failure artifacts upload, along with the screenshot and HTML snapshot of the failed test. 

### Sort results and test logs primarily by the suite name

This aligns with the test running order (test suite -> branch). See the comparison below of the current vs. new results upload content.

| performance results upload (current) | performance results upload (new) |
| ------------- | ------------- |
| Screenshot 2023-03-10 at 11 44 42 | Screenshot 2023-03-10 at 11 45 03 |

I've updated the test log to reflect the above and added the `test rounds` info that was missing:

| runner info (current) | runner info (new) |
| ------------- | ------------- |
| Screenshot 2023-03-10 at 11 47 34 | Screenshot 2023-03-10 at 11 47 55 |

> **Note**
> While working on this one, I've found out that the release performance tests have been failing for the past three months (https://github.com/WordPress/gutenberg/issues/48920)

### Write the env-specific configs directly

For the performance tests, we're creating "environment" folders for each branch we want to test against. Those folders need to contain env-specific `wp-env` config, which we're currently copying from a [boilerplate config](https://github.com/WordPress/gutenberg/blob/04b53b71e2111ab2426b3ee4a9a717a62337afa7/bin/plugin/utils/.wp-env.performance.json) that we copy over and amend with the target env config values. There's no point keeping this boilerplate and doing the extra copy step, because we can write this file directly. Also, this way we can provide absolute paths which makes it more readable.

## Testing Instructions

CI:
- Tests should pass, 
- Results should be uploaded [as described above](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Archive%20performance%20results%20for%20each%20comparison%20step) (3 files for each test - branch A raw results, branch B raw results & computed results),
- Result filenames should begin with the test name,
- Logs for the comparison* jobs should be updated [as described above](https://github.com/WordPress/gutenberg/pull/48684#:~:text=Sort%20results%20and%20test%20logs%20primarily%20by%20the%20current%20suite%20name).

Locally:
- Checkout this branch
- Make sure your local `wp-env` is stopped,
- Run `bin/plugin/cli.js perf refactor/perf-test-results-path trunk`,
- (Go for lunch or sth - it will take a while),
- The job should pass,
- All results should be saved to the `./artifacts` folder.
- The same should happen if you run tests individually with e.g. `npm run test:performance test-editor`

To test if the failure artifacts are still uploaded as expected:
- Make any of the perf specs fail by adding `expect(false).toBe(true)`,
- Run `bin/plugin/cli.js perf refactor/perf-test-results-path trunk` again,
- Once the job fail, the failure artifacts (screenshot + HTML snapshot) should be uploaded to `./artifacts` as well.

### CODE REVIEW

File: bin/plugin/commands/performance.js [hunk 1]
```diff
@@ -182,9 +182,14 @@ function curateResults( testSuite, results ) {
...
 async function runTestSuite( testSuite, performanceTestDirectory, runKey ) {
+	fs.mkdirSync( './__test-results', { recursive: true } );
+	const testResultsFilePath = path.resolve(
+		`./__test-results/${ runKey }.results.json`
+	);
+
 	try {
 		await runShellScript(
-			`npm run test:performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
+			`TEST_RESULTS_FILE_PATH="${ testResultsFilePath }" npm run test:performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
```
Comment (WunderBart) [inline bin/plugin/commands/performance.js]: Pass the file path to store directly inside the `./__test-results` folder, so we can avoid storing alongside the sourcecode in CI.

File: bin/log-performance-results.js [hunk 2]
```diff
@@ -10,22 +10,25 @@ const [ token, branch, hash, baseHash, timestamp ] = process.argv.slice( 2 );
...
 		metricsPrefix: '',
 	},
 	{
-		file: 'front-end-block-theme-performance-results.json',
+		file: 'front-end-block-theme.results.json',
 		metricsPrefix: 'block-theme-',
 	},
 	{
-		file: 'front-end-classic-theme-performance-results.json',
+		file: 'front-end-classic-theme.results.json',
```
Comment (dmsnell) [inline bin/log-performance-results.js]: is there a need to remove this? it was my assumption that we might end up with more test results in the future and this could help make it obvious which result belongs to which test.

it's fine to change it, and we can handle it later of course, but it has been helpful to me at least seeing all the information in the filename

File: bin/plugin/commands/performance.js [hunk 3]
```diff
@@ -178,42 +178,27 @@ function curateResults( testSuite, results ) {
...
-	);
-	fs.mkdirSync( './__test-results', { recursive: true } );
-	fs.copyFileSync( resultsFile, `./__test-results/${ runKey }.results.json` );
-	const rawResults = await readJSONFile(
-		path.join(
-			performanceTestDirectory,
-			`packages/e2e-tests/specs/performance/${ testSuite }.test.results.json`
-		)
+	await runShellScript(
+		`npm run test:performance -- ${ testSuite } --wordpress-artifacts-path=${ artifactsPath } --results-filename=${ resultsFilename }`,
```
Comment (dmsnell) [inline bin/plugin/commands/performance.js]: if we're relying on the `ENV` is it necessary to pass this here or could the performance test read the same `ENV`?

File: bin/plugin/commands/performance.js [hunk 4]
```diff
@@ -225,6 +210,7 @@ async function runTestSuite( testSuite, performanceTestDirectory, runKey ) {
 async function runPerformanceTests( branches, options ) {
 	const runningInCI = !! process.env.CI || !! options.ci;
 	const TEST_ROUNDS = options.rounds || 1;
+	const artifactsPath = process.env.WP_ARTIFACTS_PATH || '';
```
Comment (dmsnell) [inline bin/plugin/commands/performance.js]: to double-check, if this is empty, it will print to the current directory? or should we pass `'./'` as the default? or something like `__dirname`?

File: bin/plugin/commands/performance.js [hunk 5]
```diff
@@ -328,15 +314,39 @@ async function runPerformanceTests( branches, options ) {
...
-				path.resolve(
-					performanceTestDirectory,
-					'bin/plugin/utils/.wp-env.performance.json'
-				) +
-				' ' +
-				environmentDirectory +
-				'/.wp-env.json'
+		// Create the config file for the current env.
+		fs.writeFileSync(
+			path.join( environmentDirectory, '.wp-env.json' ),
```
Comment (dmsnell) [inline bin/plugin/commands/performance.js:331]: similar question here about removing `performance` from the name. when I download these files they lose some inherent linkage to the performance CI job that created them.

File: bin/plugin/commands/performance.js [hunk 6]
```diff
@@ -393,32 +403,39 @@ async function runPerformanceTests( branches, options ) {
...
 		/** @type {Array<Record<string, WPPerformanceResults>>} */
 		const rawResults = [];
-		// Alternate three times between branches.
 		for ( let i = 0; i < TEST_ROUNDS; i++ ) {
+			const roundInfo = `round ${ i + 1 } of ${ TEST_ROUNDS }`;
+			log( `    >> Suite: ${ testSuite } (${ roundInfo })` );
 			rawResults[ i ] = {};
 			for ( const branch of branches ) {
-				const runKey = `${ branch }_${ testSuite }_run-${ i }`;
+				const runKey = `${ testSuite }_${ branch }_run-${ i }`;
```
Comment (dmsnell) [inline bin/plugin/commands/performance.js]: is there context behind this swap? not the most important thing, but it stands out in this PR where we're working on the path changes.

I guess there's a claim that grouping by test suite is more favorable than grouping by branch?

File: bin/log-performance-results.js (hunk 2 above)
Comment (WunderBart) [reply bin/log-performance-results.js]: The idea here was that if CI zips it into `performance-results` folder, then it doesn't need a `.performance-results.js` suffix. On second thought, though, there's no zipping when running locally, and everything lands in the `artifacts` folder, so it would be better to keep that suffix AND add it to the raw result `json`s as well. Will make an update.

File: bin/plugin/commands/performance.js (hunk 3 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: Yeah, I thought it would, but it looks like only the top-level process ([`node cli.js perf`](https://github.com/WordPress/gutenberg/blob/f2e31bdd0f925614ac73be9415ba5117e53617ad/.github/workflows/performance.yml#L56)) receives the `ENV` from CI. The `await runShellScript( "npm run ...` spawns a separate process that doesn't receive the same `ENV` so I needed to figure out a way to pass it down. Maybe there's a better way to share the env, but at my current level of understanding, this made the most sense.

File: bin/plugin/commands/performance.js (hunk 4 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: If this is empty, it will be resolved down the stream (per job) when the puppeteer env is set up (when running `npm run test:performance`): 
https://github.com/WordPress/gutenberg/blob/b3ab0eed6fcd921364f4ceda36ffb55cecf94c33/packages/scripts/config/jest-environment-puppeteer/index.js#L45-L49

Note that the `GITHUB_WORKSPACE` var is not available here when running in CI (as mentioned [here](https://github.com/WordPress/gutenberg/pull/48684#discussion_r1133783175)) so it will always resolve to the `cwd`. This was causing the artifacts to be saved along the `env` folders that we [create during the perf tests setup](https://github.com/WordPress/gutenberg/blob/4538099f41f81783720bc9c6d619bf06002ce79e/bin/plugin/commands/performance.js#L284) and the necessity to copy them later on to the workspace's `./__test-results` folder, so that they're available for the CI.

File: bin/plugin/commands/performance.js (hunk 5 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js:331]: [It was the same before](https://github.com/WordPress/gutenberg/blob/699dd44767d4b77fb5499fc764e093057b01ced5/bin/plugin/commands/performance.js#L331-L340) - we were copying the `.wp-env.performance.json` and renaming it to `.wp-env.json`, so in this regard nothing has changed. I assumed the default `.wp-env.json` name is used for the `wp-env start` script to pick it up without any extra arguments.

File: bin/plugin/commands/performance.js (hunk 6 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: I've swapped it because it's the actual order in which we [iterate through those tests](https://github.com/WordPress/gutenberg/blob/4538099f41f81783720bc9c6d619bf06002ce79e/bin/plugin/commands/performance.js#L411-L442) and how we compare the results (`suite A: branch X vs. branch Y`, `suite B: branch X vs. branch Y`, etc.). So both the log info and results filenames will now be compiled in this fashion:

| runner info (current) | runner info (new) |
| ------------- | ------------- |
| Screenshot 2023-03-10 at 11 47 34 | Screenshot 2023-03-10 at 11 47 55 |

| results upload (current) | results upload (new) |
| ------------- | ------------- |
| Screenshot 2023-03-10 at 11 44 42 | Screenshot 2023-03-10 at 11 45 03 |

File: bin/log-performance-results.js (hunk 2 above)
Comment (WunderBart) [reply bin/log-performance-results.js]: https://github.com/WordPress/gutenberg/pull/48684/commits/7b1b36ce7d0dc295ca10534b1c87f53baef3c7fa

File: bin/plugin/commands/performance.js (hunk 6 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: As a side note, I think we should also replace `trunk` in the reference measurements with commit SHA, because `trunk` alone doesn't say much. Also, the summary should point to what is the reference and what is subject of the comparison, and display a 3rd column (delta) to show whether we're slowing down or speeding up, for example:

```
>> Comparing trunk (13a14ca) with the current branch (32b4c12)

┌──────────────────────┬─────────────────────────┬────────────────────────┬─────────────────┐
│       (index)        │    13a14ca (reference)  │    32b4c12 (subject)   │        Δ        │
├──────────────────────┼─────────────────────────┼────────────────────────┼─────────────────┤
│    serverResponse    │       '235.12 ms'       │      '233.46 ms'       │      -0.5%      │
│      firstPaint      │       '251.88 ms'       │      '186.88 ms'       │     +18.2%      │
│   domContentLoaded   │       '685.96 ms'       │      '651.18 ms'       │      +0.5%      │
│        loaded        │       '686.72 ms'       │      '652.04 ms'       │      x.xx%      │
│ firstContentfulPaint │       '8436.28 ms'      │       '8057 ms'        │      x.xx%      │
│      firstBlock      │       '9095.38 ms'      │      '8713.32 ms'      │      x.xx%      │
│         type         │        '42.98 ms'       │       '43.03 ms'       │      x.xx%      │
│       minType        │        '41.9 ms'        │       '41.92 ms'       │      x.xx%      │
│       maxType        │        '45.71 ms'       │       '46.52 ms'       │      x.xx%      │
│    typeContainer     │        '14.81 ms'       │       '14.99 ms'       │      x.xx%      │
│   minTypeContainer   │        '13.35 ms'       │       '13.69 ms'       │      x.xx%      │
│   maxTypeContainer   │        '18.87 ms'       │       '17.7 ms'        │      x.xx%      │
│        focus         │        '40.99 ms'       │       '44.78 ms'       │      x.xx%      │
│       minFocus       │        '35.64 ms'       │       '35.16 ms'       │      x.xx%      │
│       maxFocus       │        '64.35 ms'       │       '71.31 ms'       │      x.xx%      │
│     inserterOpen     │        '33.11 ms'       │       '32.39 ms'       │      x.xx%      │
│   minInserterOpen    │        '27.96 ms'       │       '27.73 ms'       │      x.xx%      │
│   maxInserterOpen    │        '52.56 ms'       │       '52.9 ms'        │      x.xx%      │
│    inserterSearch    │        '13.53 ms'       │       '12.29 ms'       │      x.xx%      │
│  minInserterSearch   │        '5.76 ms'        │       '6.11 ms'        │      x.xx%      │
│  maxInserterSearch   │        '19.31 ms'       │       '18.43 ms'       │      x.xx%      │
│    inserterHover     │        '26.31 ms'       │       '26.4 ms'        │      x.xx%      │
│   minInserterHover   │        '22.51 ms'       │       '22.38 ms'       │      x.xx%      │
│   maxInserterHover   │        '47.75 ms'       │       '46.44 ms'       │      x.xx%      │
│     listViewOpen     │       '157.32 ms'       │      '159.09 ms'       │      x.xx%      │
│   minListViewOpen    │       '142.89 ms'       │      '142.65 ms'       │      x.xx%      │
│   maxListViewOpen    │       '235.52 ms'       │      '253.54 ms'       │      x.xx%      │
└──────────────────────┴─────────────────────────┴────────────────────────┴─────────────────┘
```

What do you think?

File: bin/plugin/commands/performance.js (hunk 3 above)
Comment (dmsnell) [reply bin/plugin/commands/performance.js]: we can manually pass through any `ENV` we want in `runShellScript`

https://github.com/WordPress/gutenberg/blob/6434e0bc181bb1c0f5fcf48520be20f1a37afb6c/bin/plugin/lib/utils.js#L17-L26

@youknowriad would it make sense to pass `env: { ...process.env, …, ...env }` as a default? instead of `env: { …, ...env }`? would that be dangerous?

File: bin/plugin/commands/performance.js (hunk 6 above)
Comment (dmsnell) [reply bin/plugin/commands/performance.js]: > I think we should also replace trunk in the reference measurements with commit SHA

if we can display both that also has value. I think different needs use them differently. for most purposes `trunk` does a better job communicating than `13a14ca` because we recognize that when a CI job runs it's running against the latest `trunk` at the time the CI job ran. we can always lookup what `13a14ca` is, but that involves additional steps when `trunk` is pretty bare and simple.

> what is the reference and what is subject of the comparison

for most runs this matters, but the script supports multiple "branches" beyond two, and in those cases I'm not sure one is the reference. if we include this we will be making a decision that _one_ of the "branches" is the reference, likely the first one. (branches is in quotes because it's really just a `git` ref).

> and display a 3rd column (delta) to show whether we're slowing down or speeding up, for example:

no strong opinion here on what we _should_ do, but I can share my personal hesitancy with this due to the fact that we're still hitting some wild variation across tests (even with the recent fixes), and percentages always make changes look simpler than they are. I like making the easy-to-misuse things harder to use, whereas simply reporting `ms` is just that, a fact reported without any interpretation, because the interpretation is itself extremely difficult to automate.

File: bin/plugin/commands/performance.js (hunk 4 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: > Note that the `GITHUB_WORKSPACE` var is not available here when running in CI (...)

⚠️ To be clear, this is (currently) true only for the performance tests because they're run via child processes. For the E2E tests that run directly from the CI, the parent process env variables are accessible here. I've made some changes in https://github.com/WordPress/gutenberg/pull/48684/commits/a53065b9b9dd37366df0eb4c5dfd3d0ee55ade22 to hopefully streamline this bit.

File: bin/plugin/commands/performance.js (hunk 3 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js]: > we can manually pass through any ENV we want in runShellScript

Yeah I've just noticed that! 😄 Made already some changes in that regard in https://github.com/WordPress/gutenberg/commit/a53065b9b9dd37366df0eb4c5dfd3d0ee55ade22. I'm not passing the whole parent env though - just the relevant bits. I'm curious though if passing the whole thing would be safe.

File: bin/plugin/commands/performance.js [hunk 7]
```diff
@@ -339,15 +326,41 @@ async function runPerformanceTests( branches, options ) {
...
+				{
+					core: 'WordPress/WordPress',
+					plugins: [ path.join( environmentDirectory, 'plugin' ) ],
+					themes: [
+						path.join(
+							performanceTestDirectory,
+							'test/emptytheme'
+						),
+						'https://downloads.wordpress.org/theme/twentytwentyone.1.7.zip',
+						'https://downloads.wordpress.org/theme/twentytwentythree.1.0.zip',
```
Comment (dmsnell) [inline bin/plugin/commands/performance.js:342]: are these things we want to hard-code here in this seemingly independent function? I could easily imagine someone wanting to update the themes and then overlooking these lines…

**Comment by dmsnell** [comment]
Really love where you've taken this.

What do you see as continued blockers to get this in, or do you think it's ready to go?

File: bin/plugin/commands/performance.js (hunk 7 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js:342]: This is only carried over from the `.wp-env.performance.json` boilerplate file. Hard-coding those here was a solution to an issue described in https://github.com/WordPress/gutenberg/pull/49063. In my mind, it was an intermediate solution because those themes should be installed within the tests that use them so that when running tests in isolation (e.g. via `npm run test:performance theme-tests`), same themes are used as in CI or when we're running locally via the CLI. I would consider this a blocker - I plan to address this in a follow-up PR.

**Comment by WunderBart** [comment]
> What do you see as continued blockers to get this in, or do you think it's ready to go?

@dmsnell, I don't see any blockers at this point. In a follow-up PR which I'm already working on, I plan to make the builds reusable for subsequent jobs (at least locally) and parallelize them (builds) to make things faster.

File: bin/plugin/commands/performance.js (hunk 7 above)
Comment (dmsnell) [reply bin/plugin/commands/performance.js:342]: > I would consider this a blocker - I plan to address this in a follow-up PR.

if you meant "I would not consider this a blocker" then that sounds good. if not, then I'm confused 😄

**Review by dmsnell — APPROVED** [review]
Thanks for the work. I don't see any reason to block this.

File: bin/plugin/commands/performance.js (hunk 7 above)
Comment (WunderBart) [reply bin/plugin/commands/performance.js:342]: Ah, good catch! Missed the important 'not' there 😅
